### ISSUE TITLE
Predicted values are the same for 4 different test sets

### ISSUE THREAD

**Issue by DariusMargineanNicolae** [JoaquinAmatRodrigo/skforecast#268]

**Comment by DariusMargineanNicolae** [comment]
I have a LGBM Regressor model and what to predict 4 different subsets of the last 28 records (like a manual cross validation to see the behaviour of my model in time). I tried using the implementation with last_window (as an argument to the predict function) but the output it's almost the same for all 4 test sets.

My df contains 237 rows, and I fit the LGBM model on all except the last 28 (to see how it performed on 4 tests of 4 weeks with a step of 4 in between). 

Here is the implementation of the code

`forecaster = ForecasterAutoreg(
            regressor=LGBMRegressor(boosting_type='dart', n_estimators=500, num_leaves=50, learning_rate=0.01, lambda_l1=0.5, lambda_l2=0.5, max_depth=10, min_child_samples=30), lags=lags_lgbm
        )

forecaster.fit(y=df_to_process_with_lags["Qty"][:-28], exog=exog[:-28])
model_name = datetime.datetime.now().strftime("%d_%m_%Y_") + 'lightgbm_best_model_4fold_cv_lags' + str(lags_lgbm) +'.pkl'
pickle.dump(forecaster, open(model_name, 'wb'))

predictions1 = forecaster.predict(steps = 4, last_window=df_to_process_with_lags["Qty"][:-28], exog = exog[:-28])
predictions1

predictions2 = forecaster.predict(steps = 4, last_window=df_to_process_with_lags["Qty"][:-20], exog = exog[:-20])
predictions2

predictions3 = forecaster.predict(steps = 4, last_window=df_to_process_with_lags["Qty"][:-12], exog = exog[:-12])
predictions3

predictions4 = forecaster.predict(steps = 4, last_window=df_to_process_with_lags["Qty"][:-4], exog = exog[:-4])
predictions4
`
And here is the output: of the for the test sets:

                            index     pred_value
predictions1 ---->  209    19.011764
                             210    19.039493
                             211    12.855817
                             212    16.667053

                            index     pred_value
predictions2 ---->  217    19.011764
                              218    19.039493
                              219    12.855817
                              220    16.667053

                            index     pred_value
predictions3 ---->  225    21.058518
                              226    19.039493
                              227    12.855817
                              228    16.667053


                            index     pred_value
predictions4 ---->  233    19.011764
                             234    19.039493
                             235    12.855817
                             236    16.667053

Am I not using correctly the last_window parameter? Or should i use multiple fits to see how the model performed on different test sets?

**Comment by JavierEscobarOrtiz** [comment]
Hello @DariusMargineanNicolae,

With the information that you show in the post, I will say that you are using `last_window` correctly.

But I suggest using [Backtesting](https://joaquinamatrodrigo.github.io/skforecast/latest/user_guides/backtesting.html) to perform this kind of analysis, this code should do the work:

```python
# Backtest forecaster
# ==============================================================================
forecaster = ForecasterAutoreg(
                 regressor=LGBMRegressor(boosting_type='dart', n_estimators=500, num_leaves=50, learning_rate=0.01, lambda_l1=0.5, lambda_l2=0.5, max_depth=10, min_child_samples=30), 
                 lags=lags_lgbm
             )

metric, predictions_backtest = backtesting_forecaster(
                                   forecaster         = forecaster,
                                   y                  = df_to_process_with_lags["Qty"],
                                   exog               = exog,
                                   initial_train_size = len(df_to_process_with_lags["Qty"][:-28]),
                                   fixed_train_size   = False,
                                   steps              = 4,
                                   metric             = 'mean_absolute_error',
                                   refit              = False,
                                   verbose            = True
                               )
```

This backtesting **without refit**, `refit = False`, will do the following:
1. Fit the forecaster like `forecaster.fit(y=df_to_process_with_lags["Qty"][:-28], exog=exog[:-28])` since `initial_train_size = len(df_to_process_with_lags["Qty"][:-28])`.
2. Predict 4 steps
3. Then, the last window will move 4 steps ahead, like `df_to_process_with_lags["Qty"][:-24]`.

You can see then all the predictions, `predictions_backtest `, and the evaluation metric, `metric`.

You can explore another option like doing some refit. The graphs in the [documentation](https://joaquinamatrodrigo.github.io/skforecast/latest/user_guides/backtesting.html) will help you a lot. 😄

**Comment by JoaquinAmatRodrigo** [comment]
Hi @DariusMargineanNicolae 
What value of lags are you using in the forecaster? 
If it is 4, then, the result of predict() will be the same for every last_window with a length higher than 4, since only the first 4 values are used.

**Comment by JavierEscobarOrtiz** [comment]
Hello @DariusMargineanNicolae 

I think the problem is probably in the regressor, check this out:

```python
from skforecast.ForecasterAutoreg import ForecasterAutoreg
from sklearn.linear_model import LinearRegression
from lightgbm import LGBMRegressor

# Generate 50 random datapoints
np.random.seed(123)
y = pd.Series(np.random.random(50))

forecaster = ForecasterAutoreg(regressor=LinearRegression(),
                               lags=4)
forecaster.fit(y)

print(forecaster.predict(4))
print(forecaster.predict(4, last_window=y[:-4]))
print(forecaster.predict(4, last_window=y[:-12]))
```
50    0.577972
51    0.454038
52    0.497240
53    0.507692
Name: pred, dtype: float64

46    0.422583
47    0.573156
48    0.463378
49    0.509567
Name: pred, dtype: float64

38    0.578311
39    0.482056
40    0.542317
41    0.481195
Name: pred, dtype: float64


Then I repeat the same thing but with your LGBM regressor:

```python
forecaster = ForecasterAutoreg(regressor=LGBMRegressor(boosting_type='dart', n_estimators=500, num_leaves=50, learning_rate=0.01, lambda_l1=0.5, lambda_l2=0.5, max_depth=10, min_child_samples=30),
                               lags=4)
forecaster.fit(y)

print(forecaster.predict(4))
print(forecaster.predict(4, last_window=y[:-4]))
print(forecaster.predict(4, last_window=y[:-12]))
```

50    0.50722
51    0.50722
52    0.50722
53    0.50722
Name: pred, dtype: float64

46    0.50722
47    0.50722
48    0.50722
49    0.50722
Name: pred, dtype: float64

38    0.50722
39    0.50722
40    0.50722
41    0.50722
Name: pred, dtype: float64

I suggest you do the same test with your data or test your model with some random data.
